direktiv/direktiv-ui
# [DIR-191] enterprise edition

This PR adds all enterprise features. Some demo screenshots and videos are attached to the end of this PR.

- A new menu item called "Permission" (only available when `VITE_IS_ENTERPRISE` is set to true)
  - Policy
  - Groups
  - Tokens 
- Changes on the API layer
  - Api factory was updated to handle an array as a response (changed `return schema.parse({ ...parsedResponse });` to `return schema.parse(parsedResponse);
`
   - errors from the API will now be handled via an `ApiErrorSchema` that contains the response object as well as an optional json with a code and message. Every time we catch an error from the API (since errors are not predictable, the current TypeScript version always types them as `unknown`), we can parse the error against this schema and get a type safe error back. This needed to be introduced to handle permission errors, but can be handy for any backend error in the future as well.
- Handle insufficient permissions of the user. There are two part about this. Mutations and queries. When the user performs a mutation without a proper permission, it will pop up a toast message. When the user has no permissions to view something, we can not use a toast, but must do that visually in a similar way as we handle the "no results" state of a query.
  - introduced two proxy hooks for`useQuery` to handle permissions
     - `useMutationWithPermissions` check the comments in `src/api/useMutationWithPermissions.ts` for a detailed explanation
     - `useQueryWithPermissions` check the comments in `src/api/useQueryWithPermissions.ts` for a detailed explanation
- Display an error on the onboarding page when namespaces could not be fetched
- The `CopyButton` does not work [when not in a secure context](https://developer.mozilla.org/en-US/docs/Web/API/Clipboard_API) (domains other than localhost with no SSL), so I added a visual indicator when copying fails. I also had a more fancy solution with a proper tooltip (you can revert 62ed5b64ccc1198d1e38f7b5cb06e520bf155607 to see it), but that had the downside of adding a dependency to the translation function and our translation file as well. This would require us to add `react-i18next` and include the translation file to Storybook and I would like to avoid this if we can. Currently, our design folder and the rest of the app have a clear separation of concerns, which I like. 

# Use enterprise edition

Since this PR heavily relies on having an enterprise installation available somewhere, it is not that easy to test everything. You should set these environment variables in place:

```
VITE_BASE="/ui/" 
VITE_IS_ENTERPRISE="TRUE"
``` 

`VITE_BASE` will configure the app to be available under this root folder. Defining e.g. `VITE_BASE="/ui/"` will place the app in the folder `ui`. This is important, as the enterprise edition will need the /api folder and other folders like /oidc (openID connect) available. Moving the ui into a separate folder will avoid url naming collisions.

`VITE_IS_ENTERPRISE` is doing multiple things:
- It will make the menu item "Permissions" visible
- It will make the `useApiKeyHandling()` hook not care about any API key anymore, since the enterprise edition is not using them at all. The enterprise edition has its own authentication layer, one level above the UI. We can assume, that a user that can access the UI, has already authenticated themselves and has a valid cookie to talk to the api.
- it will affect the menu at the top right to
  - always show a user icon
  - always show a logout link, which is a predefined url to a route outside the UI.
- it will enable the hook useRefreshSession() to make a call every minute to check if the user is still logged in / refresh the session. When this call fails with an unauthenticated error, the user will be redirected to the logout url

# Test enterprise edition

## start enterprise with a non enterprise API

When you want to start the dev server in enterprise mode, but only have a non-enterprise backend, you need to change the following (when your backend api requires an api key)

change `src/hooksNext/useApiKeyHandling.ts`
```typescript
// set isEnterprise to false to enable API key handling
// const { VITE_IS_ENTERPRISE: isEnterprise } = env; 
const isEnterprise = false; 
```

change `src/componentsNext/AuthenticationProvider/index.tsx`
```typescript
// disable useRefreshSession
// useRefreshSession({ enabled: env.VITE_IS_ENTERPRISE }); 
```

## test the permission menu item with a non enterprise API

You can revert the commit 55d31eb4a576fb9956e4c20615007e31f7769fce to enable mocks I used for development, it mocks most of the endpoints with some static data



### How to test insufficient permissions

A quick way to check how the UI responses to an insufficient permission result from the API is to inject some code in the API factory.

```typescript

// START: INJECT ERROR
if (
  url(urlParams).startsWith("/api/namespaces/stefan/events") &&
  method === "GET"
) {
  const errorObject = await createApiErrorFromResponse(
    new Response(
      JSON.stringify({
        code: "some code",
        // message: "optional message from the serer",
      }),
      {
        status: 401,
        statusText: "Unauthorized",
      }
    )
  );

  return Promise.reject(errorObject);
}
// END: INJECT ERROR

if (res.ok) {
```

**Important:** if you inject an error in the api factory, it might be the case that this fetch request will be detected as unauthorized, but a different streaming endpoint exist and has proper permissions (because no permission error is injected). In the example above, the `isAllowed` property of `useEvents` is false at the beginning and changes to true when streamed data comes in. This is actually a correct behavior, it should never be the case that a user has no permissions to an endpoint via nonstreaming but can access it via streaming. Both methods are always bound together.

To easily disable streaming just, go to `src/api/streaming.ts` and make `useStreaming` return `null`

# Screenshots and Videos

Demo of the permissions. Mocked data are used in this demo. Not all of it works perfect and is consistent in terms of data, but it shows all the features of the app.

https://github.com/direktiv/direktiv-ui/assets/121789579/4d412bab-7072-4460-a6b9-075d172a2a3e

---

Alert on the onboarding page, when the namespaces call fails
<img width="1044" alt="CleanShot 2023-09-14 at 13 27 21@2x" src="https://github.com/direktiv/direktiv-ui/assets/121789579/80b502f7-3789-4e57-8cc3-480c6b5a62b5">

---

Page with no permissions
<img width="1393" alt="CleanShot 2023-09-15 at 08 46 49@2x" src="https://github.com/direktiv/direktiv-ui/assets/121789579/86e915d8-8d2d-4e3a-9eab-5e3850359c74">

---
CopyButton displaying an error when clipboard API is not available

https://github.com/direktiv/direktiv-ui/assets/121789579/c0e6815c-4f35-4dca-bf1b-f038a51d4688





# Follow-up tickets
- [E2E tests](https://linear.app/direktiv/issue/DIR-836/enterprise-edition)
